cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cassert>
+#include <functional>
+#include <iostream>
+#include <limits>
+
+namespace cudf {
+namespace fixed_point {
+
+template <typename T>
+struct weak_typedef {
+    T _t;
+    explicit weak_typedef(T t) : _t(t) {}
+    operator T() const { return _t; }
+};
+
+using scale_type = weak_typedef<int32_t>;
+
+enum class Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type {
+        return scale_type{-scale};
+    }
+
+    // perform this operation when constructing with positive scale
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) -> double {
```

Comment:
Thinking about this more, I think the shift amount should be factored out so it can be computed once per column rather than once per element. The scale factor will be stored in the column — the data will just be integers. So you will need a function that just computes the shift amount (or divisor). The shift will actually take the divisor as a parameter, so that it can be applied in parallel to all rows.

Reply:
This is clean up now. And factoring out shift will be in follow up job.